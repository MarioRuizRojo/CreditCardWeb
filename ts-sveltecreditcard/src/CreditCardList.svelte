<script lang="ts">
  import { _ } from "svelte-i18n";
  import { onMount } from "svelte";
  
  import type { CreditCardC } from "./CreditCardC";
  import { creditCardServiceC } from "./globals";

  /**
   * Credit cards list to show
   */
  let creditCardList: CreditCardC[] = [];
  $: {
    creditCardList = creditCardList;
  }
  /**
   * It calls the service to make a request to the server for update the credit cards list
   */
  async function getCreditCardList(): Promise<void> {
    creditCardList = await creditCardServiceC.serverGetCreditCards();
  }
  /**
   * It refreshes the credit cards list and it subscribes the updateCreditCardList function
   * to the observable list updates of the service
   */
  onMount(async function (): Promise<void> {
    creditCardList = [];
    getCreditCardList();
    creditCardServiceC.getNewList().subscribe(updateCreditCardList);
  });
  /**
   * It refreshes the shown list using the service's updated list
   * @param newList service's updated list
   */
  function updateCreditCardList(newList: CreditCardC[]): void {
    if (newList != undefined) {
      creditCardList = newList;
    }
  }
  /**
   * It updates the credit card to edit observable object of service
   * with the clicked credit card data
   * @param creditCard clicked credit card from the list
   */
  function clickEditCreditCard(creditCard: CreditCardC) {
    creditCardServiceC.addCreditCardToEdit(creditCard);
  }
  /**
   * It calls the service to make a request of removal of a credit card to the server
   * if credit card was successfully deleted then it refreshes the credit cards list
   * @param creditCard clicked credit card from the list
   */
  function clickDeleteCreditCard(creditCard: CreditCardC) {
    creditCardServiceC.serverDeleteCreditCard(creditCard.getId())
    .then(function (deleted): void {
        if (deleted) {
            getCreditCardList();
        }
    })
    .catch((error1: Error) => {
        console.error(error1.message);
    });
  }
</script>

<div class="card">
  <div class="card-body">
    <h5 class="title">{$_("creditcards")}</h5>
    <table class="table">
      <tbody>
        {#each creditCardList as creditCard, index}
          <tr>
            <td>{creditCard.getHolderName()}</td>
            <td>{creditCard.getCardNumber()}</td>
            <td>{creditCard.getExpirationDate()}</td>
            <td>
              <i
                data-testid={"edit" + index}
                on:click={() => {
                  clickEditCreditCard(creditCard);
                }}
                class="far fa-edit text-info"
              />
              <i
                data-testid={"delete" + index}
                on:click={() => {
                  clickDeleteCreditCard(creditCard);
                }}
                class="far fa-trash-alt text-danger"
              />
            </td>
          </tr>
        {/each}
      </tbody>
    </table>
  </div>
</div>

<style>
  .title {
    text-align: center;
    font-family: "Potta One", cursive;
  }
</style>
