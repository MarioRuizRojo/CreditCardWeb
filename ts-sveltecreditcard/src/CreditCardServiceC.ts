import { Observable, Subject } from 'rxjs';

import { urlAddCreditCard, urlDeleteCreditCard, urlGetCreditCards, urlUpdateCreditCard } from './constants';
import { CreditCardC } from './CreditCardC';

/**
 * Service to make request calls to the server and implement observer pattern
 * to communicate components with each other
 */
export class CreditCardServiceC {
    /**
     * observer pattern
     * to communicate CreditCardEditor with CreditCardList
     * when click in edit button
     */
    private creditCardToEdit$ = new Subject<CreditCardC>();
    /**
     * observer pattern
     * to communicate CreditCardEditor with CreditCardList
     * when click in form's submit button of the CreditCardEditor
     */
    private creditCardList$ = new Subject<CreditCardC[]>();

    //----------SERVER FUNCTIONS----------
    /**
     * Request to the server to add a credit card to the list
     * @param creditCardToSave credit card to add
     * @returns promise of the request process
     */
    public serverAddCreditCard(creditCardToSave: CreditCardC): Promise<any> {
        let promiseJSON: Promise<any>;
        if (creditCardToSave == undefined) {
            promiseJSON = Promise.reject("credit card to save cannot be empty");
        } else {
            promiseJSON = fetch(
                urlAddCreditCard,
                {
                    method: 'POST',
                    body: JSON.stringify(creditCardToSave),
                    headers: {
                        'Content-Type': 'application/json'
                    }
                }
            )
            .then(function (answer): Promise<any> {
                return answer.json();
            });
        }
        return promiseJSON;
    }

    /**
     * Request to the server to get the credit cards list
     * @returns promise of the request process with a credit cards list if it succeeds
     */
    public serverGetCreditCards(): Promise<CreditCardC[]> {
        return new Promise(function (resolve, reject): void {
            fetch(urlGetCreditCards)
            .then(function (answer): Promise<any> {
                return answer.json();
            })
            .then(function (json: any): void {
                try {
                    let creditCardList1: CreditCardC[] = [];
                    let list: any[] = json;
                    list.forEach(function (cardJson: any): void {
                        let creditCard: CreditCardC = new CreditCardC(cardJson);
                        creditCard.setId(cardJson.id);
                        creditCard.setCreationDate(cardJson.creationDate);
                        creditCard.setUpdateDate(cardJson.updateDate);
                        creditCardList1.push(creditCard);
                    });
                    resolve(creditCardList1);
                } catch (error1) {
                    reject(error1);
                }
            })
            .catch((error2: Error) => {
                console.error(error2.message);
                reject(error2);
            });
        });
    }

    /**
     * Request to the server to delete a credit card that is in the list
     * @param id identifier of the credit card to delete
     * @returns promise of the request process
     */
    public serverDeleteCreditCard(id: string): Promise<any> {
        let promiseJSON: Promise<any>;
        if (id == undefined) {
            promiseJSON = Promise.reject("id of credit card to delete cannot be empty");
        } else {
            promiseJSON = fetch(
                urlDeleteCreditCard,
                {
                    method: 'POST',
                    body: JSON.stringify({ id: id }),
                    headers: {
                        'Content-Type': 'application/json'
                    }
                }
            )
            .then(function (answer): Promise<any> {
                return answer.json();
            });
        }
        return promiseJSON;
    }

    /**
     * Request to the server to update a credit card that is in the list
     * @param creditCardToUpdate identifier and credit card new data to update
     * @returns promise of the request process
     */
    public serverUpdateCreditCard(creditCardToUpdate: CreditCardC): Promise<any> {
        let promiseJSON: Promise<any>;
        if (creditCardToUpdate == undefined) {
            promiseJSON = Promise.reject("credit card to update cannot be empty");
        }
        else {
            promiseJSON = fetch(
                urlUpdateCreditCard,
                {
                    method: 'POST',
                    body: JSON.stringify(creditCardToUpdate),
                    headers: {
                        'Content-Type': 'application/json'
                    }
                }
            )
            .then(function (answer): Promise<any> {
                return answer.json();
            });
        }
        return promiseJSON;
    }
    //-------RX observer pattern--------
    //----------RXjs FUNCTIONS----------
    //EDIT CARD
    /**
     * It is called from CreditCardList
     * 
     * CreditCardEditor is subscribed to its updates
     * 
     * When it updates CreditCardEditor
     * enter edit mode and fill form with the data of the credit 
     * card to edit
     * @param creditCard credit card to edit
     */
    public addCreditCardToEdit(creditCard: CreditCardC) {
        this.creditCardToEdit$.next(creditCard);
    }

    /**
     * CreditCardEditor uses it to subscribe to updates on creditCardToEdit$
     */
    public getCreditCardToEdit(): Observable<CreditCardC> {
        return this.creditCardToEdit$.asObservable();
    }
    //UPDATE LIST
    /**
     * It is called from CreditCardEditor when submit button is clicked
     * 
     * CreditCardList is subscribed to its updates
     * 
     * When it updates CreditCardList renders again the credit cards list
     * with the new credit cards list data
     * @param newCreditCardList new credit cards list to render
     */
    public setNewList(newCreditCardList: CreditCardC[]) {
        this.creditCardList$.next(newCreditCardList);
    }

    /**
     * CreditCardList uses it to subscribe to updates on creditCardList$
     */
    public getNewList(): Observable<CreditCardC[]> {
        return this.creditCardList$.asObservable();
    }
}


