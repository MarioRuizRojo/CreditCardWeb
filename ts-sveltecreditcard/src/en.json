{
    "addcard": "Add Credit Card",
    "loading": "Loading...",
    "placeholdercardholdername": "Card Holder Name",
    "placeholdercardsnumber": "Card's Number",
    "placeholderdate": "MM/YY",
    "placeholdercvv": "CVV",
    "add": "Add",
    "required": "Required",
    "onlyletters": "Should contain only letters",
    "should16dig": "Should contain 16 digits",
    "shoulddateform": "Should have date format",
    "should3dig": "Should contain 3 digits",
    "savechanges": "Save Changes",
    "creditcards": "Credit Cards",
    "creditcardapp": "Credit Cards App",
    "selectlang": "Select language",
    "editcard": "Edit Credit Card"
}