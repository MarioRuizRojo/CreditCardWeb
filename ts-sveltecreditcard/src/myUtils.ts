import { iconClassBase, iconClassGreen, iconClassRed } from "./constants";

/**
 * It returns red color clases if the validation fails and green ones if it succeeds
 * @param passed validation status
 * @param iconClassX css base class of icon
 * @returns css classes of icon
 */
export function iconSuccedOrFailed(passed: boolean, iconClassX: string): string {
    if (passed) {
        return iconClassBase + iconClassX + iconClassGreen;
    }
    else {
        return iconClassBase + iconClassX + iconClassRed;
    }
}

/**
 * Very useful dummy function that has no parameters and does nothing
 */
export function dummyFunction(): void {
    //do nothing
}