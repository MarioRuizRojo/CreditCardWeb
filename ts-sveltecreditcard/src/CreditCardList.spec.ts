import CreditCardList from './CreditCardList.svelte';
import { render, fireEvent } from '@testing-library/svelte';
import { Observable, Subject } from 'rxjs';

import type { CreditCardC } from './CreditCardC';
import { CreditCardServiceC } from './CreditCardServiceC';
import { twoCreditCards, oneCreditCard } from './fixtures/fixtures';
import { dummyFunction } from './myUtils';

//language support in tests
import { locale, _, addMessages } from 'svelte-i18n';
import en from './en.json';

/**
 * Mocked returned value from serverGetCreditCards, when called returns Promise like two 
 * credit cards list from test data set
 */
let serviceGivesListTwoCards: Promise<CreditCardC[]> = new Promise((resolve, reject) => resolve(twoCreditCards));
/**
 * Mocked returned value from serverGetCreditCards, when called returns Promise like one 
 * credit card list from test data set
 */
let serviceGivesListOneCard: Promise<CreditCardC[]> = new Promise((resolve, reject) => resolve(oneCreditCard));
/**
 * Mocked serverDeleteCreditCard function to check if its called
 */
let mockServiceDelete: jest.Mock<any, any>;
/**
 * Mocked addCreditCardToEdit function to check if its called
 */
let mockServiceEdit: jest.Mock<any, any>;
/**
 * Mocked serverGetCreditCards function to mock returned values
 */
let mockGetListCards: jest.Mock<any, any>;

/**
 * It sets up the language translation engine before running the tests
 */
beforeAll(function () {
    addMessages('en', en);
    locale.set('en');
});

/**
 * Mock all service functions
 */
beforeEach(function () {
    mockServiceDelete = jest.fn(function (): Promise<boolean> {
        return new Promise((resolve, reject) => resolve(true))
    });
    CreditCardServiceC.prototype.serverDeleteCreditCard = mockServiceDelete;
    mockServiceEdit = jest.fn(dummyFunction);
    CreditCardServiceC.prototype.addCreditCardToEdit = mockServiceEdit;

    //to mock service get Credit Card list function
    //not all tests use it
    mockGetListCards = jest.fn()
        .mockReturnValueOnce(serviceGivesListTwoCards)
        .mockReturnValueOnce(serviceGivesListOneCard);
});

/**
 * It expects default values for CreditCardList attributes
 */
it('check constructor empty card list', function () {
    const rendered = render(CreditCardList);
    expect(rendered).toMatchSnapshot();
});

/**
 * It expects two credit cards list rendered in CreditCardList
 */
it('check constructor 2 cards in list of cards', async function () {
    CreditCardServiceC.prototype.serverGetCreditCards = mockGetListCards;
    const rendered = render(CreditCardList);
    const deleteButton: HTMLElement = await rendered.findByTestId("delete1");//wait until onMount
    expect(rendered).toMatchSnapshot();
});

/**
 * It expects serverDeleteCreditCard to have been called
 */
it('check if onClick delete triggers a call to service delete', async function () {
    CreditCardServiceC.prototype.serverGetCreditCards = mockGetListCards;
    const { findByTestId } = render(CreditCardList);
    //this waits until there is a match, so until CreditCardList's onMount
    const deleteButton: HTMLElement = await findByTestId("delete0");
    await fireEvent.click(deleteButton);
    expect(mockServiceDelete).toHaveBeenCalledTimes(1);
});

/**
 * It renders two credit card list and deletes one
 * It expects serverDeleteCreditCard to have been called
 * and one credit card list rendered at the end
 */
it('check if onClick delete removes one credit card from the list of 2 cards', async function () {
    CreditCardServiceC.prototype.serverGetCreditCards = mockGetListCards;
    const rendered = render(CreditCardList);
    //this waits until there is a match, so until CreditCardList's onMount
    const deleteButton2ndItem: HTMLElement = await rendered.findByTestId("delete1");//delete second
    await fireEvent.click(deleteButton2ndItem);
    expect(rendered).toMatchSnapshot();
});

/**
 * It renders two credit card list and clicks edit on one of them
 * It expects addCreditCardToEdit to have been called with the first
 * credit card of the list as a parameter
 */
it('check if onClick edit triggers a call to service addCreditCardToEdit with first card object', async function () {
    CreditCardServiceC.prototype.serverGetCreditCards = mockGetListCards;
    const rendered = render(CreditCardList);
    //this waits until there is a match, so until CreditCardList's onMount
    const editButtonFirstItem: HTMLElement = await rendered.findByTestId("edit0");
    await fireEvent.click(editButtonFirstItem);
    expect(mockServiceEdit).toHaveBeenLastCalledWith(twoCreditCards[0]);
});

/**
 * It subscribes CreditCardList to local observable instance creditCardList$
 * It updates that instance to one credit card list from test data set
 * It expects one credit card list rendered in CreditCardList
 */
it('check if subscribe triggers the rendering of the new cards list', async function () {
    let creditCardList$ = new Subject<CreditCardC[]>();
    //mocks service new card list subscribe function
    const mockNewListSubscribe: jest.Mock<any, any> = jest.fn(
        function (): Observable<CreditCardC[]> {
            return creditCardList$.asObservable();
        }
    );
    CreditCardServiceC.prototype.getNewList = mockNewListSubscribe;
    const rendered = render(CreditCardList);//empty list
    creditCardList$.next(oneCreditCard);//fires subscribe with new list, fixtures list
    const deleteButton: HTMLElement = await rendered.findByTestId("delete0");//wait until there is new list
    expect(rendered).toMatchSnapshot();
});