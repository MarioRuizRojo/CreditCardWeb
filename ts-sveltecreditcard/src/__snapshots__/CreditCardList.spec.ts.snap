// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check constructor 2 cards in list of cards 1`] = `
Object {
  "component": CreditCardList {
    "$$": Object {
      "after_update": Array [],
      "before_update": Array [],
      "bound": Object {},
      "callbacks": Object {},
      "context": Map {},
      "ctx": Array [
        Array [
          CreditCardC {
            "cardNumber": "1234561234567890",
            "creationDate": undefined,
            "cvv": "123",
            "expirationDate": "11/11",
            "holderName": "holderName",
            "updateDate": undefined,
          },
          CreditCardC {
            "cardNumber": "1234567890123456",
            "creationDate": undefined,
            "cvv": "321",
            "expirationDate": "01/01",
            "holderName": "holderName2",
            "updateDate": undefined,
          },
        ],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
      ],
      "dirty": Array [
        -1,
      ],
      "fragment": Object {
        "c": [Function],
        "d": [Function],
        "i": [Function],
        "l": [Function],
        "m": [Function],
        "o": [Function],
        "p": [Function],
      },
      "not_equal": [Function],
      "on_destroy": Array [
        [Function],
        [Function],
      ],
      "on_disconnect": Array [],
      "on_mount": Array [],
      "props": Object {},
      "skip_bound": false,
      "update": [Function],
    },
    "$capture_state": [Function],
    "$inject_state": [Function],
  },
  "container": <body>
    <div>
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <h5
            class="title svelte-180ga2g"
          >
            Credit Cards
          </h5>
           
          <table
            class="table"
          >
            <tbody>
              <tr>
                <td>
                  holderName
                </td>
                 
                <td>
                  1234561234567890
                </td>
                 
                <td>
                  11/11
                </td>
                 
                <td>
                  <i
                    class="far fa-edit text-info"
                    data-testid="edit0"
                  />
                   
                  <i
                    class="far fa-trash-alt text-danger"
                    data-testid="delete0"
                  />
                </td>
                 
              </tr>
              <tr>
                <td>
                  holderName2
                </td>
                 
                <td>
                  1234567890123456
                </td>
                 
                <td>
                  01/01
                </td>
                 
                <td>
                  <i
                    class="far fa-edit text-info"
                    data-testid="edit1"
                  />
                   
                  <i
                    class="far fa-trash-alt text-danger"
                    data-testid="delete1"
                  />
                </td>
                 
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </body>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`check constructor empty card list 1`] = `
Object {
  "component": CreditCardList {
    "$$": Object {
      "after_update": Array [],
      "before_update": Array [],
      "bound": Object {},
      "callbacks": Object {},
      "context": Map {},
      "ctx": Array [
        Array [],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
      ],
      "dirty": Array [
        -1,
      ],
      "fragment": Object {
        "c": [Function],
        "d": [Function],
        "i": [Function],
        "l": [Function],
        "m": [Function],
        "o": [Function],
        "p": [Function],
      },
      "not_equal": [Function],
      "on_destroy": Array [
        [Function],
        [Function],
      ],
      "on_disconnect": Array [],
      "on_mount": Array [],
      "props": Object {},
      "skip_bound": false,
      "update": [Function],
    },
    "$capture_state": [Function],
    "$inject_state": [Function],
  },
  "container": <body>
    <div>
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <h5
            class="title svelte-180ga2g"
          >
            Credit Cards
          </h5>
           
          <table
            class="table"
          >
            <tbody />
          </table>
        </div>
      </div>
    </div>
  </body>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`check if onClick delete removes one credit card from the list of 2 cards 1`] = `
Object {
  "component": CreditCardList {
    "$$": Object {
      "after_update": Array [],
      "before_update": Array [],
      "bound": Object {},
      "callbacks": Object {},
      "context": Map {},
      "ctx": Array [
        Array [
          CreditCardC {
            "cardNumber": "1234561234567890",
            "creationDate": undefined,
            "cvv": "123",
            "expirationDate": "11/11",
            "holderName": "holderName",
            "updateDate": undefined,
          },
        ],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
      ],
      "dirty": Array [
        1,
      ],
      "fragment": Object {
        "c": [Function],
        "d": [Function],
        "i": [Function],
        "l": [Function],
        "m": [Function],
        "o": [Function],
        "p": [Function],
      },
      "not_equal": [Function],
      "on_destroy": Array [
        [Function],
        [Function],
      ],
      "on_disconnect": Array [],
      "on_mount": Array [],
      "props": Object {},
      "skip_bound": false,
      "update": [Function],
    },
    "$capture_state": [Function],
    "$inject_state": [Function],
  },
  "container": <body>
    <div>
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <h5
            class="title svelte-180ga2g"
          >
            Credit Cards
          </h5>
           
          <table
            class="table"
          >
            <tbody>
              <tr>
                <td>
                  holderName
                </td>
                 
                <td>
                  1234561234567890
                </td>
                 
                <td>
                  11/11
                </td>
                 
                <td>
                  <i
                    class="far fa-edit text-info"
                    data-testid="edit0"
                  />
                   
                  <i
                    class="far fa-trash-alt text-danger"
                    data-testid="delete0"
                  />
                </td>
                 
              </tr>
              <tr>
                <td>
                  holderName2
                </td>
                 
                <td>
                  1234567890123456
                </td>
                 
                <td>
                  01/01
                </td>
                 
                <td>
                  <i
                    class="far fa-edit text-info"
                    data-testid="edit1"
                  />
                   
                  <i
                    class="far fa-trash-alt text-danger"
                    data-testid="delete1"
                  />
                </td>
                 
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </body>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`check if subscribe triggers the rendering of the new cards list 1`] = `
Object {
  "component": CreditCardList {
    "$$": Object {
      "after_update": Array [],
      "before_update": Array [],
      "bound": Object {},
      "callbacks": Object {},
      "context": Map {},
      "ctx": Array [
        Array [
          CreditCardC {
            "cardNumber": "1234561234567890",
            "creationDate": undefined,
            "cvv": "123",
            "expirationDate": "11/11",
            "holderName": "holderName",
            "updateDate": undefined,
          },
        ],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
      ],
      "dirty": Array [
        -1,
      ],
      "fragment": Object {
        "c": [Function],
        "d": [Function],
        "i": [Function],
        "l": [Function],
        "m": [Function],
        "o": [Function],
        "p": [Function],
      },
      "not_equal": [Function],
      "on_destroy": Array [
        [Function],
        [Function],
      ],
      "on_disconnect": Array [],
      "on_mount": Array [],
      "props": Object {},
      "skip_bound": false,
      "update": [Function],
    },
    "$capture_state": [Function],
    "$inject_state": [Function],
  },
  "container": <body>
    <div>
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <h5
            class="title svelte-180ga2g"
          >
            Credit Cards
          </h5>
           
          <table
            class="table"
          >
            <tbody>
              <tr>
                <td>
                  holderName
                </td>
                 
                <td>
                  1234561234567890
                </td>
                 
                <td>
                  11/11
                </td>
                 
                <td>
                  <i
                    class="far fa-edit text-info"
                    data-testid="edit0"
                  />
                   
                  <i
                    class="far fa-trash-alt text-danger"
                    data-testid="delete0"
                  />
                </td>
                 
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </body>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
